python/lsst/daf/butler/core/datastoreCacheManager.py

```diff
@@ -223,21 +497,277 @@ def move_to_cache(self, uri: ButlerURI, ref: DatasetRef) -> Optional[ButlerURI]:
+            path = self.cache_directory.join(entry)
+
+            # If the file happens to also exist in the exempt folder
+            # skip removal since a process is actively using the file.
+            if self._temp_exempt_directory.join(entry).exists():
+                log.debug("Skipping removal of %s since it is currently in use", path)
+                continue
+
+            self._cache_entries.pop(entry)
+            log.debug("Removing file from cache: %s", path)
+            try:
+                path.remove()
+            except FileNotFoundError:
+                pass
+
+    def _expire_cache(self) -> None:
+        """Expire the files in the cache.
+
+        Notes
+        -----
+        The expiration modes are defined by the config or can be overridden.
+        Available options:
+
+        * ``files``: Number of files.
+        * ``datasets``: Number of datasets
+        * ``size``: Total size of files.
+        * ``age``: Age of files.
+
+        The first three would remove in reverse time order.
+        Number of files is complicated by the possibility of disassembled
+        composites where 10 small files can be created for each dataset.
+
+        Additionally there is a use case for an external user to explicitly
+        state the dataset refs that should be cached and then when to
+        remove them. Overriding any global configuration.
+        """
+        if self._expiration_mode is None:
+            # Expiration has been disabled.
+            return
+
+        # mypy can't be sure we have set a threshold properly
+        if self._expiration_threshold is None:
+            log.warning("Requesting cache expiry of mode %s but no threshold set in config.",
+                        self._expiration_mode)
+            return
+
+        # Sync up cache. There is no file locking involved so for a shared
+        # cache multiple processes may be racing to delete files. Deleting
+        # a file that no longer exists is not an error.
+        self.scan_cache()
+
+        if self._expiration_mode == "files":
+            n_files = len(self._cache_entries)
+            n_over = n_files - self._expiration_threshold
+            if n_over > 0:
+                sorted_keys = self._sort_cache()
+                keys_to_remove = sorted_keys[:n_over]
+                self._remove_from_cache(keys_to_remove)
+            return
+
+        if self._expiration_mode == "datasets":
+            # Count the datasets, using reverse date order
+            # so that oldest turn up first.
+            datasets = defaultdict(list)
+            for key in self._sort_cache():
+                entry = self._cache_entries[key]
+                datasets[entry.ref].append(key)
+
+            n_datasets = len(datasets)
+            n_over = n_datasets - self._expiration_threshold
+            if n_over > 0:
+                keys_to_remove = []
+                removed = 0
+                for dataset in datasets:
```

Comment:
Can't you just iterate over `datasets[:n_over]` like above, skipping the whole `removed` thing?

Reply:
Simplified.